web/services/clusters.go

```diff
@@ -48,28 +58,76 @@ func (s *clustersService) GetAll(filters map[string][]string) (models.ClusterLis
 		}
 	}
 
-	err := db.Find(&clusterList).Error
+	err := db.Find(&clusters).Error
 	if err != nil {
 		return nil, err
 	}
 
+	var clusterList models.ClusterList
+	for _, cluster := range clusters {
+		clusterList = append(clusterList, cluster.ToModel())
+	}
+
 	err = s.enrichClusterData(clusterList)
 	if err != nil {
 		return nil, err
 	}
 
-	if tagsFilter, ok := filters["tags"]; ok {
-		clusterList = filterByTags(clusterList, tagsFilter)
-	}
-
 	if healthFilter, ok := filters["health"]; ok {
 		clusterList = filterByHealth(clusterList, healthFilter)
 	}
 
-	//db.Model(&User{}).Where("name = ?", "jinzhu").Count(&count) > 0
 	return clusterList, nil
 }
 
+func (s *clustersService) GetAllClusterTypes() ([]string, error) {
+	var clusterTypes []string
+
+	err := s.db.Model(&entities.Cluster{}).
+		Distinct().
+		Pluck("cluster_type", &clusterTypes).
+		Error
+
+	if err != nil {
+		return nil, err
+	}
+
+	return clusterTypes, nil
+}
+
+func (s *clustersService) GetAllSIDs() ([]string, error) {
+	var sids pq.StringArray
+
+	err := s.db.Model(&entities.Cluster{}).
+		Where("sids IS NOT NULL").
+		Distinct().
+		Pluck("unnest(sids)", &sids).
```

Comment:
Is the `Distinct` going to be applied after `unnesting`? or before? 

Asking because I don't know how `Distinct` works with `text[]`

:innocent: :flushed:

Reply:
distinct just distinct to the select clause, while unnest expands the array in rows.. this is all pgsql magic and we don't have to worry about it.

Reply:
yeah sure :joy: thanks

Reply:
@nelsonkopliku btw i dont think that the order matters in gorm, i believe that till you don't end the chain with an  ending method (find, first, scan, pluck, etc) the condition is added to the query builder in a smart way

Reply:
makes sense. Thanks for clarifying!